File: s2/L1/cpp/primenumbers.cpp
#include "primenumbers.hpp"
#include <cmath>
#include <iostream>

template<typename T>
void print_arr(T* arr, int size){
    for(int i = 0; i < size; i++)
        std::cout << i << " : " << arr[i] << '\n';
    std::cout << "=================\n";
}

prime_numbers::prime_numbers(int n){
    bool* sieve = new bool[n + 1];
    for(int i = 2; i <= n; i++)
        sieve[i] = true;
    for(int i = 2; i <= std::sqrt(n); i++)
        if(sieve[i])
            for(int j = i * i; j <= n; j += i)
                sieve[j] = false;

    int32_t* raw = new int32_t[n + 1];
    uint16_t count = 0;
    for(int i = 2; i <= n; i++)
        if(sieve[i]){
            raw[count] = i;
            count++;
        }

    this->data = new int32_t[count];
    this->length = count;
    for(int i = 0; i < count; i++)
        this->data[i] = raw[i];

    //print_arr(sieve, n + 1);
    //print_arr(raw, n + 1);
    //print_arr(data, count);
    
    delete[] sieve;
    delete[] raw;
}

prime_numbers::~prime_numbers(){
    delete[] data;
    length = 0;
}

int prime_numbers::get_number(int n){
    if(n < 0 || n >= this->length) return -1;
    return data[n];
}

File: s2/L1/java/PrimeNumbers.java
import java.util.*;

public class PrimeNumbers{

    private void printArr(int[] arr){
        for(int i = 0; i < arr.length; i++){
            System.out.println(Integer.toString(i) + " : " + Integer.toString(arr[i]));
        }
        System.out.println("-------------------------");
    }
    private void printArr(boolean[] arr){
        for(int i = 0; i < arr.length; i++){
            System.out.println(Integer.toString(i) + " : " + Boolean.toString(arr[i]));
        }
        System.out.println("-------------------------");
    }

    PrimeNumbers(int n){
        boolean[] sieve = new boolean[n + 1];
        for(int i = 2; i <= n; i++){
            sieve[i] = true;
        }
        for(int i = 2; i <= Math.sqrt(n); i++){
            if(sieve[i]){
                for(int j = i * i; j <= n; j += i){
                    sieve[j] = false;
                }
            }
        }
        //printArr(sieve);
        int[] raw = new int[n + 1];
        int count = 0;
        //System.out.println(sieve.length);
        for(int i = 2; i <= n; i++){
            if(sieve[i]){
                raw[count] = i;
                count++;
            }
        }
        //printArr(raw);
        data = new int[count];
        for(int i = 0; i < count; i++){
            data[i] = raw[i];
        }
        //printArr(data);
    }

    public int getNumber(int m){
        if(m < 0 || m >= data.length) return -1;
        return data[m];
    }
    public int length(){
        return data.length;
    }

    private int data[];
}
